store/src/java-test/com/zimbra/cs/event/analytics/EventMetricTest.java

```diff
@@ -0,0 +1,270 @@
+        key = new MetricKey<RatioMetric, Double, RatioIncrement>(MetricType.EVENT_RATIO, globalParams);
+        EventMetric<RatioMetric, Double, RatioIncrement> globalReadRatio = metrics.getMetric(key);
+
+        assertTrue("initializer.getInitialValue() should have been triggered", initializedFlag.get() == true);
+        assertEquals("initial contact1 read ratio should be 0", new Double(0), contactReadRatio.getValue());
+        assertEquals("initial global read ratio should be 0", new Double(0), globalReadRatio.getValue());
+
+        logMsgEvents(CONTACT_1, 1, 3, EventType.SEEN);
+        logMsgEvents(CONTACT_1, 1, 2, EventType.READ);
+        logMsgEvents(CONTACT_2, 10, 3, EventType.SEEN);
+        logMsgEvents(CONTACT_2, 10, 1, EventType.READ);
+
+        assertEquals("new contact1 read ratio should be 2/3", new Double(2d/3), contactReadRatio.getValue());
+        assertEquals("new global read ratio should be 1/2", new Double(1d/2), globalReadRatio.getValue());
+    }
+
+    @Test
+    public void testTimeToOpen() throws Exception {
+        AtomicBoolean initializedFlag = new AtomicBoolean(false);
+        MetricInitializer<RatioMetric, Double, RatioIncrement> initializer = new DummyRatioInitializer(initializedFlag);
+
+        EventDifferenceParams contactParams = new EventDifferenceParams(EventType.SEEN, EventType.READ, CONTACT_1);
+        contactParams.setInitializer(initializer);
+        MetricKey<RatioMetric, Double, RatioIncrement> key = new MetricKey<RatioMetric, Double, RatioIncrement>(MetricType.TIME_DELTA, contactParams);
+        EventMetric<RatioMetric, Double, RatioIncrement> contactTTO = metrics.getMetric(key);
+
+        EventDifferenceParams globalParams = new EventDifferenceParams(EventType.SEEN, EventType.READ);
+        globalParams.setInitializer(initializer);
+        key = new MetricKey<RatioMetric, Double, RatioIncrement>(MetricType.TIME_DELTA, globalParams);
+        EventMetric<RatioMetric, Double, RatioIncrement> globalTTO = metrics.getMetric(key);
+
+        assertTrue("initializer.getInitialValue() should have been triggered", initializedFlag.get() == true);
+        assertEquals("initial contact1 TTO should be 0", new Double(0), contactTTO.getValue());
+        assertEquals("initial global TTO should be 0", new Double(0), globalTTO.getValue());
+
+        long timestamp = System.currentTimeMillis();
+        //1 second difference for contact1
+        logMsgEvents(CONTACT_1, 1, 1, EventType.SEEN, timestamp-1000);
+        logMsgEvents(CONTACT_1, 1, 1, EventType.READ, timestamp);
+
+        //2 second difference for contact2
+        logMsgEvents(CONTACT_2, 2, 1, EventType.SEEN, timestamp-2000);
+        logMsgEvents(CONTACT_2, 2, 1, EventType.READ, timestamp);
+
+        assertEquals("new contact1 TTO should be 1 second", new Double(1), contactTTO.getValue());
+        assertEquals("initial global TTO should be 1.5 seconds", new Double(1.5), globalTTO.getValue());
```

Comment:
what makes initial global TTO to change here after it was set to 0 by the initializer?

Reply:
This is a copy/paste error, it should say "new global TTO". Logging a TTO of 1  second for contact1 and 2 seconds for contact2 makes the global TTO 1.5.